src/mdns/mdns_avahi.cpp

```diff
@@ -533,79 +663,184 @@ otbrError PublisherAvahi::PublishService(const char *   aHostName,
-            it->mPort == aPort)
+        SuccessOrExit(error = ResetGroup(serviceIt->mGroup));
+    }
+    else
+    {
+        otbrLog(OTBR_LOG_INFO, "[mdns] update service %s.%s for host %s", aName, aType, logHostName);
+        avahiError = avahi_entry_group_update_service_txt_strlst(serviceIt->mGroup, AVAHI_IF_UNSPEC, mProtocol,
+                                                                 AvahiPublishFlags{}, aName, aType, mDomain, last);
+        if (avahiError == 0 && mServiceHandler != nullptr)
         {
-            otbrLog(OTBR_LOG_INFO, "MDNS update service %s", aName);
-            error = avahi_entry_group_update_service_txt_strlst(
-                mGroup, AVAHI_IF_UNSPEC, mProtocol, static_cast<AvahiPublishFlags>(0), aName, aType, mDomain, last);
-            SuccessOrExit(error);
-            ret = OTBR_ERROR_NONE;
-            ExitNow();
+            // The handler should be called even if the request can be processed synchronously
+            mServiceHandler(aName, aType, OTBR_ERROR_NONE, mServiceHandlerContext);
         }
+        ExitNow();
     }
 
-    otbrLog(OTBR_LOG_INFO, "MDNS create service %s", aName);
-    error = avahi_entry_group_add_service_strlst(mGroup, AVAHI_IF_UNSPEC, mProtocol, static_cast<AvahiPublishFlags>(0),
-                                                 aName, aType, mDomain, aHostName, aPort, last);
-    SuccessOrExit(error);
+    otbrLog(OTBR_LOG_INFO, "[mdns] create service %s.%s for host %s", aName, aType, logHostName);
+    avahiError =
+        avahi_entry_group_add_service_strlst(serviceIt->mGroup, AVAHI_IF_UNSPEC, mProtocol, AvahiPublishFlags{}, aName,
+                                             aType, mDomain, aHostName, aPort, last);
+    SuccessOrExit(avahiError);
 
-    {
-        Service service;
-        strcpy_safe(service.mName, sizeof(service.mName), aName);
-        strcpy_safe(service.mType, sizeof(service.mType), aType);
-        service.mPort = aPort;
-        mServices.push_back(service);
-    }
+    otbrLog(OTBR_LOG_INFO, "[mdns] commit service %s.%s", aName, aType);
+    avahiError = avahi_entry_group_commit(serviceIt->mGroup);
+    SuccessOrExit(avahiError);
 
-    ret = OTBR_ERROR_NONE;
+    serviceIt->mHostName = safeHostName;
+    serviceIt->mPort     = aPort;
 
 exit:
 
-    if (error)
+    if (avahiError)
+    {
+        error = OTBR_ERROR_MDNS;
+        otbrLog(OTBR_LOG_ERR, "Failed to publish service for avahi error: %s!", avahi_strerror(avahiError));
+    }
+    else if (error != OTBR_ERROR_NONE)
     {
-        ret = OTBR_ERROR_MDNS;
-        otbrLog(OTBR_LOG_ERR, "Failed to publish service for avahi error: %s!", avahi_strerror(error));
+        otbrLog(OTBR_LOG_ERR, "Failed to publish service: %s!",
+                error == OTBR_ERROR_ERRNO ? strerror(errno) : otbrErrorString(error));
     }
 
-    if (ret == OTBR_ERROR_ERRNO)
+    if (error != OTBR_ERROR_NONE && serviceIt != mServices.end())
```

Comment:
Is there a case where `avahiError != 0` and `error != OTBR_ERROR_NONE`? If not, we can combine in above else-if statement?

Reply:
See the comment above.